espnet3/collect_stats.py

```diff
@@ -1,16 +1,435 @@
+
+    if hasattr(cfg, "collate_fn") and cfg.collate_fn is not None:
+        return instantiate(cfg.collate_fn)
+    else:
+        return CommonCollateFn(int_pad_value=-1)
+
+
+def _build_dataset(config: DictConfig):
+    dataset = _instantiate_dataset(config.dataset_config, config.mode)
+    shard_idx = config.get("shard_idx")
+    if shard_idx is not None:
+        if not hasattr(dataset, "shard"):
+            raise RuntimeError("Dataset does not support sharding")
+        dataset = dataset.shard(shard_idx)
+
+    if hasattr(dataset, "use_espnet_collator"):
+        dataset.use_espnet_collator = True
+    return dataset
+
+
+def _build_model(config: DictConfig):
+    model_config = config.model_config
+    if not isinstance(model_config, DictConfig):
+        model_config = OmegaConf.create(model_config)
+    task = config.get("task")
+    if task:
+        model = get_espnet_model(task, model_config)
+    else:
+        model = instantiate(model_config)
+
+    collect_fn = getattr(model, "collect_feats", None)
+    if collect_fn is None or not callable(collect_fn):
+        raise AttributeError(
+            "Model is missing required callable 'collect_feats' method."
+        )
+    return model
+
+
+def _chunk_indices(num_items: int, batch_size: int) -> List[List[int]]:
+    if batch_size <= 0:
+        raise ValueError("batch_size must be a positive integer")
+    batches = [
+        list(range(i, min(i + batch_size, num_items)))
+        for i in range(0, num_items, batch_size)
+    ]
+    return [b for b in batches if b]
+
+
+def _instantiate_dataset(dataset_config, mode: str):
+    cfg = dataset_config
+    if not isinstance(cfg, DictConfig):
+        cfg = OmegaConf.create(cfg)
+
+    organizer = instantiate(cfg)
+    dataset = getattr(organizer, mode, None)
+    if dataset is None:
+        raise ValueError(f"Dataset organizer does not provide split '{mode}'")
+    return dataset
+
+
+def _dataset_length(dataset_config, mode: str, shard_idx: Optional[int] = None) -> int:
+    dataset = _instantiate_dataset(dataset_config, mode)
+    if shard_idx is not None:
+        if not hasattr(dataset, "shard"):
+            raise RuntimeError("Dataset does not support sharding")
+        dataset = dataset.shard(shard_idx)
+    return len(dataset)
+
+
+class CollectStatsInferenceProvider(EnvironmentProvider):
+    """EnvironmentProvider tailored for collect-stats jobs."""
+
+    def __init__(
+        self,
+        *,
+        model_config,
+        dataset_config,
+        dataloader_config,
+        mode: str,
+        task: Optional[str] = None,
+        shard_idx: Optional[int] = None,
+        params: Optional[Dict[str, Any]] = None,
+    ):
+        cfg = OmegaConf.create({})
+        cfg.model_config = model_config
+        cfg.dataset_config = dataset_config
+        cfg.dataloader_config = dataloader_config
+        cfg.mode = mode
+        cfg.task = task
+        cfg.shard_idx = shard_idx
+        cfg.update(**params)
```

Comment:
Might raise an error if `params` is None.

Reply:
Thank you, I will update to `params: Optional[Dict[str, Any]] = dict(),`